File: src/Systems/Objective-C/patterns/strategy.m
#import <Foundation/Foundation.h>

@protocol PricingStrategy <NSObject>
- (NSInteger)price:(NSInteger)value;
@end

@interface RegularPricing : NSObject <PricingStrategy>
@end

@implementation RegularPricing
- (NSInteger)price:(NSInteger)value { return value; }
@end

@interface VipPricing : NSObject <PricingStrategy>
@end

@implementation VipPricing
- (NSInteger)price:(NSInteger)value { return value * 80 / 100; }
@end

static NSInteger ApplyPrice(NSInteger value, id<PricingStrategy> strategy) {
    return [strategy price:value];
}

BOOL verifyStrategy(void) {
    id<PricingStrategy> regular = [RegularPricing new];
    id<PricingStrategy> vip = [VipPricing new];
    return ApplyPrice(100, regular) == 100 && ApplyPrice(100, vip) == 80;
}

#ifndef GENKIDAMA_STRATEGY_EMBEDDED
int main(void) {
    @autoreleasepool {
        NSCAssert(verifyStrategy(), @"Objective-C Strategy contract failed");
        puts("Objective-C Strategy: passed");
    }
    return 0;
}
#endif

File: src/Web/Dart/patterns/strategy.dart
typedef PricingStrategy = int Function(int value);

int price(int value, PricingStrategy strategy) => strategy(value);

int regular(int value) => value;
int vip(int value) => value * 80 ~/ 100;

void verifyStrategy() {
  if (price(100, regular) != 100 || price(100, vip) != 80) {
    throw StateError('Strategy contract failed');
  }
}

void main() {
  verifyStrategy();
  print('regular=100;vip=80');
}
